app/javascript/mastodon/reducers/compose.js

```diff
@@ -259,10 +259,15 @@ export default function compose(state = initialState, action) {
   case COMPOSE_UPLOAD_PROGRESS:
     return state.set('progress', Math.round((action.loaded / action.total) * 100));
   case COMPOSE_MENTION:
-    return state
-      .update('text', text => `${text}@${action.account.get('acct')} `)
-      .set('focusDate', new Date())
-      .set('idempotencyKey', uuid());
+    return state.withMutations(map => {
+      map.update('text', text => [text.trim(), `@${action.account.get('acct')} `].filter((str) => str.length !== 0).join(' '));
+      map.set('focusDate', new Date());
+      map.set('idempotencyKey', uuid());
+
+      if (map.get('privacy') === 'direct') {
```

Comment:
Don't do this. Sometimes I reply to multiple people and instead of typing their names, I would use the dropdown option to mention them. This does not mess with the type of message I'm composing. If it turned my DM into a public message without me noticing/expecting it that would be really bad.

Reply:
Hence the reason I said it might be contentious — wasn't sure if that behaviour was desired.

I'll remove this change, but keep the one for the correct insertion of mentions.